### DISCUSSION TITLE
Accessibility "Style" Guidelines

### DISCUSSION THREAD

**Discussion by ShrutiVellanki** [fgpv-vpgf/fgpv-vpgf#4104] [Team Posts]
## Accessibility and Accessibility Testing Concerns (Overview)

Accessibility issues were brought up in today's (March 7/2019) RAMP Project Meeting. There was confusion and concern over:
- The time-consuming aspect of accessibility testing
- How to maintain accessibility when existing panels are migrated to Panels API
- How to maintain a user friendly experience by: 
  1. Keeping navigation style consistent across apps (when to use tab keys vs escape keys vs arrow keys to navigate)
  2. Keeping navigation non-tedious (for example by creating larger "component groups" that hold smaller "sub components" for items such as table headers)

## The Proposal

1. Written Accessibility "Style" Guidelines for RAMP Components (I will address this below)
2.  PR-by-PR accessibility testing (addressed in a separate issue over [here](https://github.com/orgs/fgpv-vpgf/teams/ramp-core/discussions/4))

### 1. Drop Down Menus
 - tabbing down a list with `Tab` key vs scrolling down a list with `up/down arrow` keys

### 2. "Component Groups" and "Sub Component" Navigation
- When to use `Arrow`, `Tab` and `Esc` keys?
- e.g. In the `enhancedTable` relevant for column header navigation
-  e.g. In the `Legend` relevant for between-layer vs within-layer navigation  

### 3. Hover text accessibility 
 - make sure page text that can't be read with a screen reader is accounted for in HTML

### 4.  Cross Map Navigation
-  Will tabbing go counter clockwise in map? Clockwise? 
-  How would a person adding a new component to a map ensure that their component follows the established tabbing order of the map?

## Style Guide's Target Audience  
- Core developers. Useful when:
  - Creating a new feature
  - Migrating existing features
  - Accessibility Testing
- Plugin authors. Useful for creating plugins that will be consistent with the rest of RAMP.

**Comment by mweech** [comment]
RAMP is complex enough and with so many people doing UI work now, it's implemented in a segregated enough fashion that it was only normal to get the divergence in approaches.  Initially routing all our major UI work through Alex for implementation/review was plugging this hole.   

This kind of accessibility style guide is a great solution and could serve as a template for a guide other larger teams can benefit from on their projects.  It provides some uniformity but also helps a new developer to understand how to approach accessibility in general.

**Comment by spencerwahl** [comment]
### Cross Map Navigation

Guidelines from WCAG: 
>focusable components receive focus in an order that preserves meaning and operability.


>When the sequence in which content is presented affects its meaning, a correct reading sequence can be programmatically determined.

To me this points to setting tab order to mimic reading a newspaper or comic book. In other words left to right, top to bottom, panel by panel.

Taking just the default ramp setup the order would be:
1. App bar + Main panel (legend, details...) since they're visually attached
2. Side panel (settings, metadata) or Table
3. Overview map
4. Bottom right controls (zoom, fullscreen, scalebar etc.)


And a 5 second idea for new components.... if we use the rule above we can take the top-left point of a panel and figure out where it should be in the tab order to a certain degree. i.e Higher things first, going left to right within a certain height buffer.

**Comment by AleksueiR** [comment]
## On Keyboard navigation

Here's an overview how the VSCode team implements keyboard accessibility in the app (which is excellent btw). Since stealing is caring, I think we can draw inspiration from them, and maybe borrow some technical details as well (it's all HTML/JS anyway).

> You can use the `Tab` key to jump between VS Code UI controls. Use `Shift+Tab` to tab in reverse order. As you tab through the UI controls, you can see an indicator around the UI element once the element gains focus.

The focus travels left to right, top to bottom, as @spencerwahl remarked on above.

In VSCode it's possible to trap your focus inside the editor since `Tab` inserts a Tab character. They use the `Ctrl+M` shortcut to change the mode and let the user out of the editor. I don't think RAMP has anything similar, but we can probably use a kindred approach to escape when tabbing through the table cells.

There seems to be roughly four different scenarios when tabbing through an application. We might find more, they the following should cover most of the interactions.

You can think about any application as a list of things. There is a top-level list of major component and each component can have a list of items inside it with each item possibly containing a sublist of additional interactive elements.

### 1. Tabbing through top-level components

This is simple - just use `Tab` to move through the top-level components of the app. The side bar is always open and all its buttons seem to be considered top-level focus targets.

In RAMP it would mean always tabbing through the main application bar.

![](https://i.imgur.com/qBnqa14.gif)

### 2. Tabbing through components and their hidden controls

Collapsible components have two focus targets - the header/toggle and body. In the collapsed state the body target is skipped. To open a component, focus on the header and press `Enter`.

Notice how after opening a component, additional hidden controls are show in the header and you need to tab through them to get to the body of the component. 

If a component is not collapsible, you don't need to include the header in the tab order.

In RAMP, this would correspond to the body of the layers or details panel or any other panel which can be collapsed using a toggle.

![](https://i.imgur.com/YC1pnsW.gif)

### 3. Tabbing through a component's body

To get to the elements in a component (in this case a list of files), you press an `Up` or `Down` navigation key. This moves focus from the component body to the first element in the list inside the component. You can move up and down the list using just the navigation keys.

Pressing `Tab` or `Shift+Tab` will move the focus from to the component header or an interactive element following the component body. The selected component list element will be remembered, and it will be focused on instead of the component body the next time you tab to the it. 

Pressing `Esc`, deselect the element and moves the focus back to the component body.

In RAMP, this would correspond to moving through the layer list where legend items are navigated to using the `Up/Down` arrow keys.

![](https://i.imgur.com/tc4mwqy.gif)

### 4. Tabbing though list item controls inside a component

In the example below, each file in the list has individual controls. You navigate to an element in the list using the arrow keys, and then you use `Tab` to navigate through the controls of the item (use `Enter` to interact). After interaction, the focus seem to move to the element following the component.

If you tab past the last control inside the item, the focus moves to the element following the component; tabbing back, focuses on the control you just left.

If, while tabbing through the item's controls, you press an arrow key, the focus will move to the next/previous element.

In RAMP, this would correspond with navigating through the legend item controls in the layers panel.

![](https://i.imgur.com/YwQExkf.gif)

**Comment by mweech** [comment]
> 
> focusable components receive focus in an order that preserves meaning and operability.
> 
> When the sequence in which content is presented affects its meaning, a correct reading sequence can be programmatically determined.
> 
> To me this points to setting tab order to mimic reading a newspaper or comic book. In other words left to right, top to bottom, panel by panel.
> 

Feel like we should incorporate points from WCAG like we are trying to comply with and declare them in our accessibility style guide so that influencing factors in our design are referenced (links to guidelines, restate key ones, etc.)

> Taking just the default ramp setup the order would be:
> 
> App bar + Main panel (legend, details...) since they're visually attached
> Side panel (settings, metadata) or Table
> Overview map
> Bottom right controls (zoom, fullscreen, scalebar etc.)
> And a 5 second idea for new components.... if we use the rule above we can take the top-left point of a panel and figure out where it should be in the tab order to a certain degree. i.e Higher things first, going left to right within a certain height buffer.

This makes sense to me..

**Comment by AleksueiR** [comment]
# Focus management 

The user can perform two, top-level interactions: tab through the top level elements and navigation through the list and tree components.

Tabbing through the top-level components will be achieved through the DOM tab order without any programmatic shifting of the focus. To achieve this, panels would need to be inserted at the proper places in the DOM. It should be possible to figure out where to include the panel by checking its positioning relative to the other panels (left-top corner coordinates), so the focus will move left to right, top to bottom.

Navigation through the list and tree components will be done without using the tab order altogether.

## Tab Order

### Description

There are two major problems need solving when considering tab order:

- moving the focus from the trigger element to the generated content (an opening panel)
- moving the focus from the generated content back to the trigger element (a closing panel)

The current solution to the first problem is to create a "bridge/link" between the trigger element and the target panel and then manually set the focus to that panel when the trigger element is clicked. These links are specified through the component markup. The moving focus back follows the same process but in reverse.

#### Proposed solution

Instead of "pushing" the focus to the opened panel, the panel itself will be responsible for "pulling" the focus in and setting it to an appropriate element. Info panels, with no interactive elements (not even the close button) don't need to do that.

When a panel is open, it is provided with a reference to the trigger (or the last focused) element and it's responsibility of the panel to set the focus to this element when the panel is closed.

### Details

#### Focus-Intern

`Focus-Intern` is a simple service keeping track of the focus history (it subscribes to the `focus` events on the map container). Right now, I think only the currently focused, focusable, and persistent element need to be tracked. 

Angular Material dropdown menus do not behave as Focus Lists - their menu items are focusable elements and need to be excluded from being tracked by the `Focus-Intern` since menu items are not permanently present in the DOM.

This can be done by adding a special attribute to the menu container. The `Focus-Intern` checks if the focused element is descendant from the marked container and discards it if it is.

#### Focus-Buddy

`Focus-Buddy` is a Panel mixin which handles `pulling` in the focus on the `open` event and restoring focus to the trigger element (or the last focused) on the `close` event. When the panel opens, it receives a reference to the trigger (or the last focused) element from the `Focus-Intern` to restore the focus to.

If the element no longer exists when the panel is closed, the focus will be set to the root of the application.

#### Panel API Modification

A small modification is needed for the Panel API to handle reopening of the panels after the overlaying panel is closed. If a panel is to be reopened:

- it is inserted into the DOM before the closing panel is removed from the DOM

- it should not "pull" in the focus; instead, the focus will be set by the closing panel

### Some Cases

#### Case 1

Opening a panel:

- the user tabs to a dropdown menu trigger and selects an item from the menu
- the panel is opened 
- focus set to the first interactive panel element
- the user close the panel 
- focus is moved to the dropdown menu trigger

#### Case 2

Opening a panel on top of another panel:

- the user opens Panel A by clicking on a button
- focus is set to Panel A
- the user clicks a button inside the Panel A which opens the Panel B on top of the Panel A
- focus is set to Panel B
- Panel A is closed since Panel B is overlaying it
- the user closes Panel B
- before Panel B is removed from the DOM, it checks if any panel can be reopen and reopens Panel A
- focus is set to Panel A which is underneath Panel B
- Panel B is closed

#### Case 3

Opening two panels and closing the first:

- the user opens Panel A
- focus is set to Panel A
- the user clicks a button inside the Panel A which opens the Panel B beside the Panel A
- focus is set to Panel B
- the user tabs back to Panel A and closes it
- focus is set to Panel B
- the user closes panel B
- the focus is moved to the application root element (since Panel A no longer exists, focus cannot be set there)

## Focus Lists and Trees

### Description

List and tree components handle focus by actually not allowing direct children to be focusable. This sounds strange, but is actually very effective.

The body of the list component is made focusable using the `tabindex=0` attribute to include it in the default tab order. Using `Tab` to move the focus forward, would focus on the next interactive element after this list component.

All list components track which item in the list is selected, if any. On the example below, none of the elements are selected and the focus ring is drawn around the list component only.

![](https://i.imgur.com/7Rvw258.png)

Using arrow keys to navigate through the list does not move the focus from the list body; instead, it internally tracks which list element is selected and highlights it in a as to appear to have focus, also removing the focus ring from the list component. The body of the list uses `-activedescendant` aria attribute to provide information about the selected element to the screen reader.

![](https://i.imgur.com/IGTjYfN.png)

This approach solves the problem of losing focus when the element is removed from the DOM. If the list item is removed from the list, the focus does not move. It stays on the list body which is again has the focus ring because no elements are selected.

### Implementation Details

[to follow]

#  

#### List navigation POC

- the list is highlighted when tabbed to (focused)
- up/down arrow keys navigation list items
- tab key moves focus through button inside a list item
- 🗑 deletes a list item and the focus goes to the body; next tab will focus on the element immediatelly after the list; shift + tab will focus on the list itself
https://jsfiddle.net/ouroborosDragon/dae1h5nb/

**Thank you @alyec and @spencerwahl for your help with brainstorming this!**
